anndata/_config.py

```diff
@@ -0,0 +1,332 @@
+            super().__setattr__(option, val)
+        elif option not in self._registered_options:
+            raise KeyError(
+                f"{option} is not an available option for anndata.\
+                Please open an issue if you believe this is a mistake."
+            )
+        else:
+            registered_option = self._registered_options[option]
+            registered_option.validator(val)
+            self._config[option] = val
+
+    def __getattr__(self, option: str) -> object:
+        """
+        Gets the option's value.
+
+        Parameters
+        ----------
+        option
+            Option to be got.
+
+        Returns
+        -------
+        object
+            Value of the option.
+        """
+        if option in self._deprecated_options:
+            deprecated = self._deprecated_options[option]
+            raise DeprecationWarning(
+                f"{option} will be removed in {deprecated.removal_version}. "
+                + deprecated.message
+            )
+        if option in self._config:
+            return self._config[option]
+        raise AttributeError(f"{option} not found.")
+
+    def __dir__(self) -> Iterable[str]:
+        return sorted(super().__dir__() + list(self._config.keys()))
+
+    def reset(self, option: Iterable[str] | str) -> None:
+        """
+        Resets option(s) to its (their) default value.
+
+        Parameters
+        ----------
+        option
+            The option(s) to be reset.
+        """
+        if isinstance(option, Iterable) and not isinstance(option, str):
+            for opt in option:
+                self.reset(opt)
+        else:
+            self._config[option] = self._registered_options[option].default_value
+
+    @contextmanager
+    def override(self, **overrides):
+        """
+        Provides local override via keyword arguments.
+
+        Parameters
+        ----------
+
+        Yields
+        ------
+        None
+        """
+        restore = {a: getattr(self, a) for a in overrides}
+        try:
+            for attr, value in overrides.items():
+                setattr(self, attr, value)
+            yield None
+        finally:
+            for attr, value in restore.items():
+                setattr(self, attr, value)
+
+    @property
+    def __doc__(self):
+        options_description = self.describe(print_description=False)
+        return self.__doc_tmpl__.format(
+            options_description=options_description,
+        )
+
+
+settings = SettingsManager()
+
+##################################################################################
+# PLACE REGISTERED SETTINGS HERE SO THEY CAN BE PICKED UP FOR DOCSTRING CREATION #
+##################################################################################
+
+##################################################################################
+##################################################################################
+
+
+def check_and_get_environ_var(
```

Comment:
what’s that for? I don’t see it being used in the other code

Reply:
This was in advance of using environment variables (i.e., for retaining categories).  I can just move it to the next PR.  It's tested and I was using for prototyping, that's why I did it here.

Reply:
Removed.